content/blog/2019-10-08-remote-read-meets-streaming.md

```diff
@@ -0,0 +1,313 @@
+client a piece of the response in a form of few chunks for the single series.
+Because protobuf has no native delimiting logic, we [`extended`](https://github.com/prometheus/prometheus/pull/5703/files#diff-7bdb1c90d5a59fc5ead16457e6d2b038R44)
+proto definition to allow sending **set of small protocol buffer messages** instead of a single, huge one. We called this mode `STREAMED_XOR_CHUNKS` remote read.
+This means that Prometheus does not need to buffer the whole response anymore. Instead it can work on each series sequentially and send a single frame per
+each `SeriesSet.Next` or batch of `SeriesIterator.Next` iterations, potentially reusing the same memory pages for next series!
+
+Now, the response of `STREAMED_XOR_CHUNKS` remote read is a set of Protobuf messages (frames) as presented below:
+
+```
+// ChunkedReadResponse is a response when response_type equals STREAMED_XOR_CHUNKS.
+// We strictly stream full series after series, optionally split by time. This means that a single frame can contain
+// partition of the single series, but once a new series is started to be streamed it means that no more chunks will
+// be sent for previous one.
+message ChunkedReadResponse {
+  repeated prometheus.ChunkedSeries chunked_series = 1;
+}
+
+// ChunkedSeries represents single, encoded time series.
+message ChunkedSeries {
+  // Labels should be sorted.
+  repeated Label labels = 1 [(gogoproto.nullable) = false];
+  // Chunks will be in start time order and may overlap.
+  repeated Chunk chunks = 2 [(gogoproto.nullable) = false];
+}
+```
+
+As you can see the frame does not include raw samples anymore. That's the second improvement we did: We send in the message
+samples batched in chunks (see [this video](https://www.youtube.com/watch?v=b_pEevMAC3I) to learn more about chunks), 
+which are exactly the same chunks we store in the TSDB. 
+
+We ended up with the following server algorithm:
+
+1. Parse request.
+1. Select metrics from TSDB.
+1. For all series:
+    * For all samples:
+        * Encode into chunks
+            * if the frame is >= 1MB; break
+  * Marshal `ChunkedReadResponse` message.
+  * Snappy compress
+  * Send the message   
+
+You can find full design [here](https://docs.google.com/document/d/1JqrU3NjM9HoGLSTPYOvR217f5HBKBiJTqikEB9UiJL0/edit#).
+
+## Benchmarks
+
+How does this performance of this new approach compare to the old solution?
+
+Let's compare remote read characteristics between Prometheus `2.12.0` and `2.13.0`. As the initial results presented 
+at the beginning of this article, I was using Prometheus as a server, and Thanos sidecar as a client of the remote read.
+I was invoking testing remote read request by running gRPC call against Thanos sidecar using `grpcurl`.
+Test was performed from my laptop (Lenovo X1 16GB, i7 8th) with Kubernetes in docker (using [kind](https://github.com/kubernetes-sigs/kind)).
```

Comment:
why did you need to use k8s?

Reply:
* Easier to run sidecars
* Easier to run init containers in right order
* Easier to limt resources
* Easier to run in cloud if my laptop is not enough
* Easier to allow running it by others like you (: or someone with Mac etc  Portability.